src/Microsoft.SqlTools.ServiceLayer/QueryExecution/DataStorage/SaveAsExcelFileStreamWriterHelper.cs

```diff
@@ -0,0 +1,583 @@
+                double excelDate;
+                if (ticks < TicksPerDay) //date empty, time only
+                {
+                    style = Style.Time;
+                    excelDate = ((double)ticks) / (double)TicksPerDay;
+                }
+                else if (ticks < ExcelDateCutoffTick) //before excel cut-off, use string
+                {
+                    AddCell(dateTime.ToString("yyyy-MM-dd", System.Globalization.CultureInfo.InvariantCulture));
+                    return;
+                }
+                else
+                {
+                    if (ticks % TicksPerDay == 0) //time empty, date only
+                    {
+                        style = Style.Date;
+                    }
+                    excelDate = ((double)(ticks - ExcelEpochTick)) / (double)TicksPerDay;
+                }
+                AddCellDateTimeInternal(excelDate, style);
+            }
+
+            /// <summary>
+            /// Write a object cell
+            /// </summary>
+            /// The program will try to output number/datetime, otherwise, call the ToString 
+            /// <param name="o"></param>
+            public void AddCell(object o)
+            {
+                if (o == null)
+                {
+                    AddCellEmpty();
+                    return;
+                }
+                switch (Type.GetTypeCode(o.GetType()))
+                {
+                    case TypeCode.Byte:
+                    case TypeCode.SByte:
+                    case TypeCode.UInt16:
+                    case TypeCode.UInt32:
+                    case TypeCode.UInt64:
+                    case TypeCode.Int16:
+                    case TypeCode.Int32:
+                    case TypeCode.Int64:
+                    case TypeCode.Decimal:
+                    case TypeCode.Double:
+                    case TypeCode.Single:
+                        AddCellBoxedNumber(o);
+                        return;
+                    case TypeCode.DateTime:
+                        AddCell((DateTime)o);
+                        return;
+                    default:
+                        AddCell(o.ToString());
+                        return;
+                }
+            }
+
+            private void AddCellBoxedNumber(object number)
+            {
+                _referenceManager.AssureColumnReference();
+
+                _writer.WriteStartElement("c");
+
+                _referenceManager.WriteAndIncreaseColumnReference();
+
+                _writer.WriteStartElement("v");
+                _writer.WriteValue(number);
+                _writer.WriteEndElement();
+
+                _writer.WriteEndElement();
+            }
+
+            // The excel epoch is 1/1/1900, but it has 1/0/1900 and 2/29/1900
+            // which is equal to set the epoch back two days to 12/30/1899
+            // new DateTime(1899,12,30).Ticks
+            private const long ExcelEpochTick = 599264352000000000L;
```

Comment:
Our coding standard defines files be laid out like this:
1. constants
2. member variables
3. constructors
4. properties *(don't ask me why 😛)*
5. public methods
6. private methods